document/template.go

```diff
@@ -0,0 +1,88 @@
+package document
+
+import (
+	"embed"
+	"fmt"
+	"io"
+	"text/template"
+)
+
+//go:embed resources/*
+var resources embed.FS
+
+// TemplateKind helps us to select where to find the template. It can either be embedded or on the host filesystem
+type TemplateKind int
+
+const ( // iota is reset to 0
+	FS   TemplateKind = iota
+	FSYS              // fsys is used for embedded templates
+)
+
+type TemplateConfig struct {
+	kind TemplateKind
+	path string
+}
+
+func NewTemplateConfig() *TemplateConfig {
+	return &TemplateConfig{
+		kind: FSYS,
+		path: "resources/document.md",
+	}
+}
+
+type RenderDocumentOption func(*TemplateConfig)
+
+// WithTemplate is a functional option to override the documentation template
+func WithTemplate(tpl string) RenderDocumentOption {
+	return func(c *TemplateConfig) {
+		c.kind = FS
+		c.path = tpl
+	}
+}
+
+// RenderDocument takes a slice of Section and generate the markdown documentation either using the default
+// embedded template or the user provided template
+func RenderDocument(out io.Writer, s []Section, opts ...RenderDocumentOption) error {
+	var tpl = NewTemplateConfig()
+
+	// Apply all the functional options to the template configurations
+	for _, opt := range opts {
+		opt(tpl)
+	}
+
+	err := renderTemplate(tpl, s, out)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func renderTemplate(tpl *TemplateConfig, args interface{}, out io.Writer) error {
```

Comment:
Could you help understand the motivation behind leaving args as `interface{}` instead of declaring the type concretely (args []Section)

Reply:
I don't see any reason to constrain the type here, the underlying `t.Execute` accepts an `interface{}`. The way I see it, this is a wrapper function around `template.Template` that handles either the template is embedded or from the host.